### PR TITLE
Refactor game start date

### LINKED ISSUE #47338: Corn vanishes, or goes from rotten to old or normal.
# Describe the bug
I collected some corn from a silo. Some of which turned rotten, so I disposed it outside. After a good night sleep it got edible again.

# Steps To Reproduce

1. get rotten corn.
2. wait or go to sleep
3. it gets either old or "normal" or it vanishes

# Expected behavior
Stay rotten.

# Screenshots
grafik
grafik

# Versions and configuration
- OS: Windows
    - OS Version: 10.0 2004
- Game Version: 0.E-9473-g83b0843-dirty [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    meins [mine],
    Nonperishable Overhaul [Nonperishable_Overhaul]
]

# Additional context
It is very hard to discribe, but the corn changes its behaviour from load to load. If I get more info, I keep you informed.

### LINKED ISSUE #50312: Initial Day (-1) ignores Spawn Delay setting -- Monsters evolve and Food Rots
### Describe the bug

In the world settings, setting Initial Day: -1 creates a world where the spawn delay seems to be anchored to Spring Day 60. Spawning before Spring Day 60 means that monsters will be highly evolved and a majority of the food rotten. Spawning in Fall or Winter means significantly evolved monsters and significant rot.

### Steps To Reproduce

Enter world settings
Set Initial Day: -1
Confirm Spawn Delay: 0
Create world, with any other settings.

Additional comparison:
Compare world generation--rot and monster evolution.
Set Initial Day to match the day previous spawned.
Compare
Set Initial Day to 60, with spawn delay to match the passage of time
Compare

### Expected behavior

With setting the initial day as random, I should be able to enter a world the day that Cataclysm happens, where food is not rotten and there are almost zero high threat zombies.

### Screenshots

_No response_

### Versions and configuration

- OS: Twister OS 2.2 (ARM-Based Debian-Linux)
- Game Version: 0.F-987-g8b32ca79de
- Graphics version: Tiles
- Ingame language: English
- Mods loaded: DDA, No npc food, no fungi

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Hirmuolio**
#### Summary

Infrastructure "Refactor scenario starting date"

#### Purpose of change

Fixes #47338
Fixes #50312

Problems in old system:
Day of cataclysm and day of game start use different random values in different places.
Day of cataclysm and day of game start in scenario are can conflict.

Most notably starting game with random date hasapproximately 50% chance to result in broken world that starts before the cataclysm happened.

This is more complete fix than #47352.
If this one is merged first then  #47352 can be closed.

#### Describe the solution

Two distinct dates: 
`calendar::start_of_cataclysm` is the day of cataclysm (sometimes referred as initial day in code).
`calendar::start_of_game` is when game starts (sometimes referred as start day in code).

User defines:
`INITIAL_DAY` = Day on which the cataclysm happens. Limited to occur during the first year.
`INITIAL_TIME` = Time (hour) on which the game starts (0-23).
`SPAWN_DELAY` = How many days after cataclysm the game starts (0-999. Limited to first year)
`SEASON_LENGTH` = Length of season (14-127).

Scenario can optionally define `"custom_initial_date"` that can contain any of the following:
`"hour"` = Time (hour) on which game starts. Defaults to 8.
`"day"` = Day of season on which game starts. Defaults to 1.
`"season"` = Season on which game starts. Defaults to SPRING.
`"year"` = Year on which game starts. Defaults to 1.

Changes:
* `INITIAL_DAY` is limited to first year.
* `"custom_initial_date"` can randomize values with -1 (it already did, doc was just wrong).
* If scenario has  `"custom_initial_date"` then it will not  default to user defined values (it actually never did).
* `SPAWN_DELAY` does not apply if the scenario defines start date.
* `INITIAL_DAY` does not directly affect scenario date.
* If `INITIAL_DAY` pushes start of cataclysm to after start of game defined by scenario then the game start is moved forwards by one year. This keeps the time of year same. 
* * This can happen for example if the scenario sets the game to start on "spring day 1 year 1" but the cataclysm is on "summer day 1 year 1". In this case the game start would be moved to "spring day 1 year 2".
* * Probably there should be some kind of warning to the player so they aren't confused why the game starts on "wrong" year. In the end it is user fault. The problem is that user has set date of cataclysm and date of start that can't easily co-exist.

User configurable season length can cause unexpect things. `"day"` in `"custom_initial_date"` may fall out of its season if user has defined very short season lengths. For example "day 20 of spring" with season length of 10 would be "day 10 of summer".
Season length agnostic date system is needed. Maybe just a system with 360 day year that is scaled and rounded.
But at least this won't create broken worlds anymore.

#### Testing

Unit tests written.
The randomness of dates is tested by generating 10 dates and checking that at least one of them is different. There is extremely small chance that all 10 randomly generated dates are the same date so the test can fail even if everything works fine. The chance for this is astronomically low.

#### Additional context

`INITIAL_TIME` is a bit poor name since it is not connected to `INITIAL_DAY` in any way. `START_TIME` would be better. But I am not sure how destructive changing the name of a setting is.

Same for `custom_initial_date`. It does changes the start date.

### CODE REVIEW

**Comment by PaulBlay** [comment]
>  In the end it is user fault. The problem is that user has set date of cataclysm and date of start that can't easily co-exist.

Maybe the user wants to play Cataclysm Dark Days Ahead before the Cataclysm? You'd need to mod all the zombies and mutants into NPCs and normal animals. It would be like Office Work Simulator (And later things go boom).

**Comment by Hirmuolio** [comment]
Do you think ability to make scenario to start on random day/hour/year is feature worth keeping?

**Comment by PaulBlay** [comment]
I suppose you could set "random start day" to be "cataclysm day" + 1 + rand(season length in days * 4) (if not eternal season) or 
"cataclysm day" + 1 + rand(91 * 4) (if is eternal season).

**Comment by Hirmuolio** [comment]
What if the start of cataclysm is in that season? Lets say it is on 45th day with 90 day seasons.
Should the whole season be valid for starting day? Then there would be 50% chance that the start would be delayed to next year.
Or should it be limited to days 46-90?

Personally I don't see the point of random scenario times.
If someone creates a scenario with start date I would assume they would be aiming at some specific date.
If they didn't care what the start date was they would leave it for user to decide and would not include it in the scenario.

**Comment by PaulBlay** [comment]
If it's next year, it's next year. There's no point in having it be more than a year from the cataclysm, though.

I think the only point of random start dates is for increases surprise value and variety. Although that's largely provided just by having  random scenario as some scenarios start at different seasons.

**Comment by actual-nh** [comment]
I'm thinking #47352 should go into 0.F, and this one `0.F Feature Freeze` (I note it's a draft currently)?

**Comment by actual-nh** [comment]
With regard to eternal seasons, I would simply have it work as if the default season length, when it's absolutely **necessary** to take seasons into account (e.g., for year length).

**Comment by actual-nh** [comment]
You state that SPAWN_DELAY (time after Cataclysm to spawn) will be overriden by a scenario specifying a start date "for the game". Does this mean that such a scenario must assume the default Cataclysm date, and its specifications are only for how long afterward the player enters the world?

**Comment by Hirmuolio** [comment]
Yes scenarios do not have the ability to change when the cataclysm happens. Only when the game starts.

The scenario date is exact day on which the game starts.
If the scenario says the game starts on day 1 of winter on year 1 then that is when the game will start.
The exception is if the map settings set the cataclysm to happen later than that date. Then the start of  game is moved to same day on next year.

**Comment by Hirmuolio** [comment]
This should be pretty much ready now.

Just needs some testing to make sure game starts right. And maybe a fresh pair of eyes to spot any weird edge cases.

**Comment by actual-nh** [comment]
Something seems to be wrong with start_date_test.cpp.

File: tests/start_date_test.cpp [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    calendar::start_of_cataclysm = calendar::turn_zero;
+    calendar::turn = calendar::turn_zero;
+}
+
+TEST_CASE( "Random scenario dates" )
+{
+    // Days counted from start of year
+    time_duration first_day_of_summer = calendar::season_length();
+    time_duration last_day_of_summer = 2 * calendar::season_length() - 1_days;
+    time_duration default_start_hour = 1_hours * 8
```
Comment (actual-nh) [inline tests/start_date_test.cpp:138]:
```suggestion
    time_duration default_start_hour = 1_hours * 8;
```

File: tests/start_date_test.cpp [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        set_scenario( scenario::generic() );
+        CHECK( get_option<int>( "INITIAL_DAY" ) == default_initial_day );
+        CHECK( get_option<int>( "INITIAL_TIME" ) == default_initial_time );
+        CHECK( get_option<int>( "SPAWN_DELAY" ) == default_spawn_delay );
+
+        g->start_calendar();
+
+        REQUIRE( calendar::start_of_cataclysm == calendar::turn_zero + 1_days * default_initial_day );
+        REQUIRE( calendar::start_of_game == calendar::turn_zero + 1_days * default_initial_day + 1_hours *
+                 default_initial_time );
```
Comment (actual-nh) [inline tests/start_date_test.cpp:27]:
```suggestion
        REQUIRE( get_option<int>( "INITIAL_DAY" ) == default_initial_day );
        REQUIRE( get_option<int>( "INITIAL_TIME" ) == default_initial_time );
        REQUIRE( get_option<int>( "SPAWN_DELAY" ) == default_spawn_delay );

        g->start_calendar();

        CHECK( calendar::start_of_cataclysm == calendar::turn_zero + 1_days * default_initial_day );
        CHECK( calendar::start_of_game == calendar::turn_zero + 1_days * default_initial_day + 1_hours *
                 default_initial_time );
```

File: tests/start_date_test.cpp [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+        set_scenario( scenario::generic() );
+        CHECK( get_option<int>( "INITIAL_DAY" ) == 100 );
+        CHECK( get_option<int>( "INITIAL_TIME" ) == 13 );
+        CHECK( get_option<int>( "SPAWN_DELAY" ) == 10 );
+
+        g->start_calendar();
+
+        REQUIRE( calendar::start_of_cataclysm == calendar::turn_zero + 1_days * 100 );
+        REQUIRE( calendar::start_of_game == calendar::turn_zero + 1_days * ( 100 + 10 ) + 1_hours * 13 );
```
Comment (actual-nh) [inline tests/start_date_test.cpp:43]:
```suggestion
        REQUIRE( get_option<int>( "INITIAL_DAY" ) == 100 );
        REQUIRE( get_option<int>( "INITIAL_TIME" ) == 13 );
        REQUIRE( get_option<int>( "SPAWN_DELAY" ) == 10 );

        g->start_calendar();

        CHECK( calendar::start_of_cataclysm == calendar::turn_zero + 1_days * 100 );
        CHECK( calendar::start_of_game == calendar::turn_zero + 1_days * ( 100 + 10 ) + 1_hours * 13 );
```

File: tests/start_date_test.cpp [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        override_option spawn_delay( "SPAWN_DELAY", "10" );
+
+        CHECK( get_option<int>( "INITIAL_DAY" ) == default_initial_day );
+        CHECK( get_option<int>( "SEASON_LENGTH" ) == default_season_length );
+
+        g->start_calendar();
+
+        REQUIRE( calendar::start_of_cataclysm == calendar::turn_zero + 1_days * default_initial_day );
+        REQUIRE( calendar::start_of_game == calendar::turn_zero + 1_days * 273 + 1_hours *
+                 default_initial_time );
```
Comment (actual-nh) [inline tests/start_date_test.cpp:62]:
```suggestion
        REQUIRE( get_option<int>( "INITIAL_DAY" ) == default_initial_day );
        REQUIRE( get_option<int>( "SEASON_LENGTH" ) == default_season_length );

        g->start_calendar();

        CHECK( calendar::start_of_cataclysm == calendar::turn_zero + 1_days * default_initial_day );
        CHECK( calendar::start_of_game == calendar::turn_zero + 1_days * 273 + 1_hours *
                 default_initial_time );
```

**Review by actual-nh — COMMENTED** [review]
A few things:
- I unfortunately don't have time at the moment to patch all of them, but while both will generate an error if they don't work out to true, REQUIRE is used for things that, if they don't work right, the test can't work right; CHECK is used for the main part of the test.
- Regarding season length, there is a specific way to check to make sure it's the default season length; see calendar_test.cpp for examples of its use, IIRC.
- Thank you for remembering to reset things back to the original!

**Comment by actual-nh** [comment]
Is there anything preventing this from being mergeable? (You can also put in "Fixes #50312", BTW.)

**Comment by Hirmuolio** [comment]
As far as I can remember it should be ready. Unless some new conflicts have risen from somewhere since the last rebase.

**Comment by actual-nh** [comment]
Ping: @jbytheway - any comments?

**Pull request merged**